src/instruction.rs

```diff
@@ -301,6 +304,17 @@ impl fmt::Display for MemoryReference {
     }
 }
 
+impl FromStr for MemoryReference {
+    type Err = SyntaxError<Self>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let tokens = lex(s)?;
+        let (_, memory_reference) = all_consuming(parse_memory_reference_with_brackets)(&tokens)
```

Comment:
The [quil spec](https://quil-lang.github.io) says in section 6.4 that a memory reference can be a lone identifier. Why are we (if I understand right) requiring the bracket syntax here instead of allowing the lone identifier?

Reply:
You're right - I had misread the source I was trying to de-duplicate in [qcs-sdk-rust/crates/lib/src/qpu/mod.rs: organize_ro_sources](https://github.com/rigetti/qcs-sdk-rust/blob/0a14eea26bbb2f4acdd323d5f56a8162ade41d8d/crates/lib/src/qpu/mod.rs#L133-L149) in that I had missed the part where a missing index was an implied `[0]`, and thought that this particular usage required the index.